Remove use of np.std in utils.py to avoid numerical errors fo…

dpanici opened the pull request:
…r large arrays
Resolves #358

f0uriest commented:
Would a better solution be to avoid use of `std` altogether? IE, for `isalmostequal` we could do `np.allclose(x, x[0])` and for `islinspaced` could do `np.allclose(np.diff(x), np.diff(x)[0])`

ddudt commented on tests/test_utils.py:
```diff
@@ -0,0 +1,35 @@
+def test_islinspaced():
+    """Test that islinspaced function works on large arrays."""
+    grid_small = LinearGrid(rho=1, M=1, N=10)
+    zeta_vals = np.unique(grid_small.nodes[:, 2])
```
can use `zeta_vals = grid.nodes[grid.unique_zeta_idx, 2]`

ddudt reviewed (changes requested):
I agree with Rory that the better solution would be to avoid calling `np.std` altogether.

dpanici commented:
addressed the comments. Took me way too long to get a good solution for using `np.allclose` for when the axis may be something other than -1... numpy indexing and broadcasting can be confusing

f0uriest commented on desc/utils.py:
```diff
@@ -271,7 +271,7 @@ def issorted(x, axis=None, tol=1e-12):
 
 
-def isalmostequal(x, axis=-1, tol=1e-12):
+def isalmostequal(x, axis=-1, rtol=1e-5, atol=1e-8):
```
should probably tighten these tolerances, otherwise seems fine

ddudt commented on tests/test_utils.py:
```diff
@@ -9,27 +9,47 @@
+    assert islinspaced(zz)
+
+    # 0D arrays will return True
+    assert isalmostequal(np.array(0))
```
this should be `assert islinspaced(np.array(0))`

ddudt commented on desc/utils.py:
```diff
@@ -271,7 +271,7 @@ def issorted(x, axis=None, tol=1e-12):
 
 
-def isalmostequal(x, axis=-1, tol=1e-12):
+def isalmostequal(x, axis=-1, rtol=1e-6, atol=1e-12):
```
need to add `rtol` and `atol` to documentation

ddudt reviewed (changes requested):
This solution seems overly complicated, but it works.

dpanici commented:
> This solution seems overly complicated, but it works.

yea without np.std I could not figure out how to get the same functionality as cleanly using np.allclose (namely the ability to have the axis of comparison be arbitrary)
